### PR TITLE
fix: prevent duplicate push metrics and deep link handling

### PR DESCRIPTION
**Pull request by levibostian**
Merges into: https://github.com/customerio/customerio-ios/pull/403

Prevent duplicate push metrics and deep link handling for push click handling.

### CODE REVIEW

File: Sources/MessagingPush/Store/PushHistory.swift [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+protocol PushHistory: AutoMockable {
+    func hasHandledPushClick(deliveryId: String) -> Bool
+    func handledPushClick(deliveryId: String)
+}
+
+// sourcery: InjectRegister = "PushHistory"
+class PushHistoryImpl: PushHistory {
+    private let keyValueStorage: KeyValueStorage
+
+    var numberOfPushesToTrack = 100
```
Comment (ami-aman) [inline Sources/MessagingPush/Store/PushHistory.swift:13]: What does this value represent? I think adding a comment would add value to it.

File: Sources/MessagingPush/Store/PushHistory.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        var clickHistory = lastPushesClicked
+
+        if clickHistory.count >= numberOfPushesToTrack {
+            // Remove oldest push click from history.
+            clickHistory = Array(clickHistory.dropFirst())
+        }
+
+        clickHistory.append(deliveryId)
+
+        lastPushesClicked = clickHistory
```
Comment (ami-aman) [inline Sources/MessagingPush/Store/PushHistory.swift:45]: Where can I find use case of this method `handledPushClick`? Is it in some other PR ?

**Review by ami-aman — APPROVED** [review]

File: Sources/MessagingPush/Store/PushHistory.swift (hunk 2 above)
Comment (levibostian) [reply Sources/MessagingPush/Store/PushHistory.swift:45]: [It's used in this PR](https://github.com/customerio/customerio-ios/pull/405/files#diff-96954b50d5a6c450f707a82cae92b37ae877dfe98b15d2006d254cae04213a4dR123-R127)

File: Sources/MessagingPush/PushClickHandler.swift [hunk 3]
```diff
@@ -118,7 +120,11 @@ class PushClickHandlerImpl: NSObject, PushClickHandler, UNUserNotificationCenter
     }
 
     private func pushClicked(_ response: UNNotificationResponse, parsedPush: CustomerIOParsedPushPayload) {
-        // TODO: prevent duplicate push metrics and deep link handling.
+        guard !pushHistory.hasHandledPushClick(deliveryId: parsedPush.deliveryId) else {
+            // push has already been handled. exit early
+            return
+        }
+        pushHistory.handledPushClick(deliveryId: parsedPush.deliveryId)
```
Comment (mrehan27) [inline Sources/MessagingPush/PushClickHandler.swift:127]: Did we write any tests for `pushClicked` anywhere? If not, can we write some?

File: Sources/MessagingPush/Store/PushHistory.swift (hunk 1 above)
Comment (levibostian) [reply Sources/MessagingPush/Store/PushHistory.swift:13]: Thanks for asking. I changed the name to `maxSizeOfHistory` to be more descriptive of it's purpose.

File: Sources/MessagingPush/Store/PushHistory.swift [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import CioInternalCommon
+import Foundation
+
+protocol PushHistory: AutoMockable {
+    func hasHandledPushClick(deliveryId: String) -> Bool
+    func handledPushClick(deliveryId: String)
+}
+
+// sourcery: InjectRegister = "PushHistory"
+class PushHistoryImpl: PushHistory {
```
[the remark this answers is not in the document]
Comment (levibostian) [reply Sources/MessagingPush/Store/PushHistory.swift:10]: Great idea. Done.

File: Sources/MessagingPush/PushClickHandler.swift (hunk 3 above)
Comment (levibostian) [reply Sources/MessagingPush/PushClickHandler.swift:137]: This feature has been heavily reliant on QA testing thus far. Automated tests around swizzling and UserNotifications has been difficult to write. I have not yet found an easy or quick way to make them. 

I do agree and think we can write some for this function after QA concludes, yes.

Comment (levibostian) [reply Sources/MessagingPush/PushClickHandler.swift:137]: I am going to merge this PR to make maintenance and QA testing easier. I made a note to add these tests and codecov in the target PR should help to keep track of these missing tests.

**Pull request merged**